### PR TITLE
[RFR] Refactor datagrid pagination

### PR DESCRIPTION
**Pull request by jpetitcolas**
- [x] Make pagination directive isolated from $location
- [x] Add link to first and last page
- [x] Use default Bootstrap markup for pagination
- [x] Add some tests on datagrid pagination directive
- [x] Fix pagination issue when too many pages (see #296)
- [x] Separate infinite pagination and normal pagination into 2 directives
- [x] Move click on page link action to the caller (listController) and pass it as a parameter

Here is how the pagination nav looks for a 9 page list:

```
at page 1:   1 2 … 9 Next »
at page 2:   « Prev 1 2 3 … 9 Next »
at page 3:   « Prev 1 2 3 4 … 9 Next »
at page 4:   « Prev 1 2 3 4 5 … 9 Next »
at page 5:   « Prev 1 … 4 5 6 … 9 Next »
at page 6:   « Prev 1 … 5 6 7 8 9 Next »
at page 7:   « Prev 1 … 6 7 8 9 Next »
at page 8:   « Prev 1 … 7 8 9 Next »
at page 9:   « Prev 1 … 8 9
```

### CODE REVIEW

File: src/javascripts/test/unit/Crud/list/maDatagridPaginationSpec.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                it('should just display that no records have been found if no item fetched', function() {
+                    scope.totalItems = 0;
+                    var element = $compile(directiveUsage)(scope)[0];
+                    scope.$digest();
+
+                    var totalElement = element.querySelector('.grid-detail .total');
+                    expect(totalElement.innerText.trim()).toBe('No record found.');
+                });
+
+                it('should display item range displayed', function() {
```
Comment (jpetitcolas) [inline src/javascripts/test/unit/Crud/list/maDatagridPaginationSpec.js]: Review description label

Comment (fzaninotto) [inline src/javascripts/test/unit/Crud/list/maDatagridPaginationSpec.js]: "it should display the content range currently displayed"

**Comment by fzaninotto** [comment]
Missing:
* it should display prev page except on the first one
* it should display next page except on the last one

File: src/javascripts/test/unit/Crud/list/maDatagridPaginationSpec.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            it("should display truncated centered page range if too many pages compared to given page range", function() {
+            });
+        });
+
+        describe('with infinite pagination', function() {
+            beforeEach(function() {
+                scope.infinite = true;
+            });
+
+            it('should not be displayed if pagination is infinite', function() {
```
Comment (fzaninotto) [inline src/javascripts/test/unit/Crud/list/maDatagridPaginationSpec.js]: Well, that's a problem in fact. It's not related with this PR, but infinite pagination hides the total number of results. We need to solve this in the UI.

File: src/javascripts/test/unit/Crud/list/maDatagridPaginationSpec.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                });
+            });
+
+            it("should not display pagination links if all items can be displayed on current page", function () {
+            });
+
+            it("should display all links if number of pages is below page range", function() {
+            });
+
+            it("should display truncated centered page range if too many pages compared to given page range", function() {
```
Comment (fzaninotto) [inline src/javascripts/test/unit/Crud/list/maDatagridPaginationSpec.js]: We haven't defined what we want, in fact, so this should probably be expanded. 

I believe we should end up displaying 5 page links at most, with something like:

                 current page
                     V
    <Prev  1  ... 45 46 47 .. 163 Next >

**Comment by fzaninotto** [comment]
Any progress?

**Comment by fzaninotto** [comment]
Switching to RFR

File: src/javascripts/ng-admin/Crud/list/DatagridPagination.html [hunk 4]
```diff
@@ -1,12 +0,0 @@
-<div ng-if="!paginationCtrl.infinite">
```
Comment (fzaninotto) [inline src/javascripts/ng-admin/Crud/list/DatagridPagination.html]: renamed to `maDatagridPagination`

File: src/javascripts/ng-admin/Crud/list/DatagridPaginationController.js [hunk 5]
```diff
@@ -1,98 +0,0 @@
-/*global define*/
```
Comment (fzaninotto) [inline src/javascripts/ng-admin/Crud/list/DatagridPaginationController.js]: renamed to `maDatagridPaginationController`

**Pull request merged**